src/account/address.cpp

```diff
@@ -1,36 +1,56 @@
 #include "account/address.h"
-#include <sodium.h>
 #include "bits.h"
 #include "bech32/bech32.h"
+#include "params.h"
+
+#include <sodium.h>
+#include <stdexcept>
 
 Address::Address(bytes const &publicKey) :
         m_pk(publicKey),
         m_bech32Address()
-{}
+{
+    if (publicKey.size() != crypto_sign_PUBLICKEYBYTES)
+        throw std::length_error(ERROR_MSG_KEY_BYTES_SIZE);
+}
 
 Address::Address(std::string const &bech32Address) :
         m_pk(),
         m_bech32Address(bech32Address)
-{}
+{
+    if (util::bech32::decode(bech32Address).first != hrp)
+        throw std::invalid_argument(ERROR_MSG_BECH32);
+}
 
 bytes Address::getPublicKey() const
 {
-    return m_pk;
+    return (m_pk.empty()) ?
+        (computePkFromBech32()) :(m_pk);
 }
 
 std::string Address::getBech32Address() const
 {
     return (m_bech32Address.empty()) ?
-           (computeBech32Address()) : (m_bech32Address);
+        (computeBech32Address()) : (m_bech32Address);
 }
 
 std::string Address::computeBech32Address() const
 {
     unsigned char pk[crypto_sign_PUBLICKEYBYTES];
     std::copy(m_pk.begin(), m_pk.end(), pk);
 
-    int const kNoBitsInByte = 8;
-    int const kNoBitsInBech32 = 5;
     return util::bech32::encode(hrp, util::convertBits(pk, crypto_sign_PUBLICKEYBYTES, kNoBitsInByte, kNoBitsInBech32, true));
 }
 
+bytes Address::computePkFromBech32() const
+{
+    bytes pk5BitsPerByte = util::bech32::decode(m_bech32Address).second;
+    unsigned int const pk5BitsSize = pk5BitsPerByte.size();
+
+    auto pkCharPtr = new unsigned char[pk5BitsSize];
+    std::copy(pk5BitsPerByte.begin(), pk5BitsPerByte.end(), pkCharPtr);
+
+    delete[] pkCharPtr;
```

Comment:
I'm a bit out of the loop with this - is it all right to call the `delete` operator, but use `pkCharPtr` afterwards - is the delete deferred to the end of the scope or something like that?

Also, why the copy of `pk5BitsPerByte` to `pkCharPtr`? Might [data()](https://www.cplusplus.com/reference/vector/vector/data/) and [this cast](https://stackoverflow.com/a/4254644/1475331) help perhaps (not sure, in this case)?

Reply:
> I'm a bit out of the loop with this - is it all right to call the delete operator, but use pkCharPtr afterwards - is the delete deferred to the end of the scope or something like that?

No, it is definitely not ok, that was a terrible mistake I've made :D . I've found the mistake while working on [this ticket](https://elrondnetwork.atlassian.net/browse/EN-10130) and fixed it there. However, I'll fix it here as well.

> Also, why the copy of pk5BitsPerByte to pkCharPtr? Might data() and this cast help perhaps (not sure, in this case)?

I've also seen that cast, but didn't try it, since I've used `std::copy` everywhere else. I've tried it now and it works fine, so I'll stick with your suggestion. However, in this [ticket](https://elrondnetwork.atlassian.net/browse/EN-10130) I will wrap/change `convertBits` function to take as parameter `bytes` instead of `unsigned char*`, so I don't have to always make that cast from outside